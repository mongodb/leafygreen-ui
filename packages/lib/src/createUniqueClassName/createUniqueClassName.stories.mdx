{/*lib.stories.mdx */}
import { Meta } from '@storybook/addon-docs';
import createUniqueClassName from '.';

<Meta title="Lib Utilities/createUniqueClassName" component={<></>} />

## createUniqueClassName

`createUniqueClassName` creates a string intended to be used as a pseudo-unique className. 
It returns a randomized 8-letter string prefixed by `lg-ui`, followed by a custom second prefix defined by the `prefix` parameter.

For example, `createUniqueClassName()` returns: 
<code>{createUniqueClassName()}</code>

With prefix `ComponentName`, `createUniqueClassName('ComponentName')` returns: 
<code>{createUniqueClassName('ComponentName')}</code>

#### Why does this exist?

Emotion no longer supports using the return value from `css` as a selector within other styles, which limits the ability to use things like well-scoped sibling and child selectors. The original use of selectors generated outside of emotion to workaround this limitation was actually given to us by the maintainers of Emotion themselves, as they know this is a limitation of the library.
Traditionally weâ€™ve used statically-named `data-` attributes to use within emotion css declarations. The intent here is to move to more performant selectors (though admittedly, that was a marginal consideration here, as are the performance benefits) that are less prone to accidental duplication.

#### How it should be used

```
  import { createUniqueClassName } from '@leafygreen-ui/lib';

  const checkClassName = createUniqueClassName();
  const inputClassName = createUniqueClassName();

  ...

  const labelHoverStyle = css`
    &:hover:not(:focus-within)
      > .${inputClassName}:not([disabled])
        + .${checkClassName} {
      box-shadow: 0 0 0 3px ${palette.gray.light2};
    }
  `;

  ...

  return (
    ...
    <input
      className={cx(inputClassName, className)}
    />
  );

```


The above component's render function will result in the following HTML:

```
<input class="lg-ui-31ed9431" />
```