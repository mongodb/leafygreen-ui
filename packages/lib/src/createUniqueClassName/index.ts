/**
 * `generateUniqueClassName` creates a string intended to be used as a pseudo-unique className.
 *
 * It returns a consecutive 4-number string prefixed by `lg-ui`, followed by a custom second prefix defined by the `prefix` param.
 *
 * @param prefix: a custom prefix that follows the `lg-ui` prefix.
 */
const generateUniqueClassName = (prefix: string) => {
  const count = classNameRegistry.get(prefix)?.length ?? 0;
  const postfix = count.toString().padStart(4, '0');
  return `lg-ui${prefix ? `-${prefix}` : ''}-${postfix}`;
};

const classNameRegistry: Map<string, Array<string>> = new Map();

const createUniqueClassName = (prefix: string = ''): string => {
  const uniqueClassName = generateUniqueClassName(prefix);

  if (classNameRegistry.has(prefix)) {
    const prefixClasses = classNameRegistry.get(prefix);
    classNameRegistry.set(prefix, [
      uniqueClassName,
      ...(prefixClasses as Array<string>),
    ]);
  } else {
    classNameRegistry.set(prefix, [uniqueClassName]);
  }

  return uniqueClassName;
};

export default createUniqueClassName;
