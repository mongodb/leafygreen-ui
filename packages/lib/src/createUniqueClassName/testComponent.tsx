import React from 'react';

import createUniqueClassName from '.';

const className1 = createUniqueClassName();
const className2 = createUniqueClassName();

export const TestComponent = ({ ...props }) => {
  return (
    <div data-testid="el-1" className={className1}>
      <div data-testid="el-2" className={className2}></div>
    </div>
  );
};
