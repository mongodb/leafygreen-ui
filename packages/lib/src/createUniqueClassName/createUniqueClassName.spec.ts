import React from 'react';
import { getByTestId, render } from '@testing-library/react';

import { TestComponent } from './testComponent';
import createUniqueClassName from '.';

const renderTestComponent = () => {
  return render(React.createElement(TestComponent));
};

describe('packages/lib/createUniqueClassName', () => {
  describe('createUniqueClassName', () => {
    test('Format is correct', () => {
      const lgPrefix = 'lg-ui-';
      const expectedUuidLength = 8;
      const result = createUniqueClassName();
      expect(result.startsWith(lgPrefix)).toBe(true);
      // eslint-disable-next-line jest/valid-expect
      expect(result.length == lgPrefix.length + expectedUuidLength);
    });

    test('Two generated classNames are not equal.', () => {
      const res1 = createUniqueClassName();
      const res2 = createUniqueClassName();
      expect(res1.valueOf()).not.toEqual(res2.valueOf());
    });

    test('Prefixes are correctly applied.', () => {
      const lgPrefix = 'lg-ui-';
      const customPrefix = 'test';
      const res1 = createUniqueClassName(customPrefix);
      const res2 = createUniqueClassName(customPrefix);
      expect(res1.startsWith(lgPrefix + customPrefix + '-')).toBe(true);
      expect(res2.startsWith(lgPrefix + customPrefix + '-')).toBe(true);
      expect(res1.valueOf()).not.toEqual(res2.valueOf());
    });
  });

  describe('rendering', () => {
    test('Separate classNames in a component are unique', () => {
      const renderedTestComponent = renderTestComponent();
      const element1 = renderedTestComponent.getByTestId('el-1');
      const element2 = renderedTestComponent.getByTestId('el-2');
      const className1 = element1.classList.value;
      const className2 = element2.classList.value;
      expect(className1).not.toEqual(className2);
    });

    test('Classnames persist on re-render', () => {
      const renderedTestComponent = renderTestComponent();
      const element1 = renderedTestComponent.getByTestId('el-1');
      const element2 = renderedTestComponent.getByTestId('el-2');
      const className1 = element1.classList.value;
      const className2 = element2.classList.value;
      renderedTestComponent.rerender(React.createElement(TestComponent));
      expect(element1.classList.value).toEqual(className1);
      expect(element2.classList.value).toEqual(className2);
    });

    test('Classnames persist when dynamically imported', async () => {
      const { TestComponent: TestComponentA } = await import('./testComponent');
      const { container: containerA } = render(
        React.createElement(TestComponentA),
      );
      const elementA1 = getByTestId(containerA, 'el-1');
      const elementA2 = getByTestId(containerA, 'el-2');

      const { TestComponent: TestComponentB } = await import('./testComponent');
      const { container: containerB } = render(
        React.createElement(TestComponentB),
      );
      const elementB1 = getByTestId(containerB, 'el-1');
      const elementB2 = getByTestId(containerB, 'el-2');

      expect(elementA1.classList.value).toEqual(elementB1.classList.value);
      expect(elementA2.classList.value).toEqual(elementB2.classList.value);
    });

    test('Classnames persist in snapshot tests', () => {
      const { container } = render(React.createElement(TestComponent));
      expect(container).toMatchSnapshot();
    });
  });
});
