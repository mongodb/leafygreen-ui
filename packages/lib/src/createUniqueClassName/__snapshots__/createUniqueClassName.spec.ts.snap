// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`packages/lib/createUniqueClassName rendering Classnames persist in snapshot tests 1`] = `
<div>
  <div
    class="lg-ui-0000"
    data-testid="el-1"
  >
    <div
      class="lg-ui-0001"
      data-testid="el-2"
    />
  </div>
</div>
`;
